packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m

```diff
@@ -543,16 +562,25 @@ - (void)setPlaybackSpeed:(double)speed {
 }
 
 - (CVPixelBufferRef)copyPixelBuffer {
+  // Ensure video sampling at regular intervals. This function is not called at exact time intervals
+  // so CACurrentMediaTime returns irregular timestamps which causes missed video frames. The range
+  // outside of which targetTime is reset should be narrow enough to make possible lag as small as
+  // possible and at the same time wide enough to avoid too frequent resets which would lead to
+  // irregular sampling. Ideally there would be a targetTimestamp of display link used by flutter
+  // engine (FlutterTexture can provide timestamp and duration or timestamp and targetTimestamp).
+  CFTimeInterval currentTime = CACurrentMediaTime();
+  if (fabs(self.targetTime - currentTime) > self.frameUpdater.duration / 2) {
+    self.targetTime = currentTime;
+  }
+  self.targetTime += self.frameUpdater.duration;
+
   CVPixelBufferRef buffer = NULL;
-  CMTime outputItemTime = [_videoOutput itemTimeForHostTime:CACurrentMediaTime()];
-  if ([_videoOutput hasNewPixelBufferForItemTime:outputItemTime]) {
-    buffer = [_videoOutput copyPixelBufferForItemTime:outputItemTime itemTimeForDisplay:NULL];
-  } else {
-    // If the current time isn't available yet, use the time that was checked when informing the
-    // engine that a frame was available (if any).
-    CMTime lastAvailableTime = self.frameUpdater.lastKnownAvailableTime;
-    if (CMTIME_IS_VALID(lastAvailableTime)) {
-      buffer = [_videoOutput copyPixelBufferForItemTime:lastAvailableTime itemTimeForDisplay:NULL];
+  CMTime outputItemTime = [self.videoOutput itemTimeForHostTime:self.targetTime];
+  if ([self.videoOutput hasNewPixelBufferForItemTime:outputItemTime]) {
+    buffer = [self.videoOutput copyPixelBufferForItemTime:outputItemTime itemTimeForDisplay:NULL];
+    if (buffer) {
+      CVBufferRelease(self.latestPixelBuffer);
+      self.latestPixelBuffer = buffer;
```

Comment:
It is.

Reply:
I think we need to retain the buffer here (rather than at the end of the function when you return). Otherwise may leak (retaining multiple times)

Reply:
> I think we need to retain the buffer here (rather than at the end of the function when you return). Otherwise may leak (retaining multiple times)

See https://github.com/flutter/packages/pull/7466#discussion_r1829941466. The code is correct, it just needs more comments inline explaining why.

Reply:
> It is.

Oops, I'm not sure how I missed that; I thought it searched the whole file for release calls.